crates/core/src/client/mod.rs

```diff
@@ -68,38 +72,68 @@ impl<P: Provider + Send + Sync + 'static> KakarotClient<P> {
-            Arc::clone(&starknet_provider),
-            kakarot_address,
             proxy_account_class_hash,
             externally_owned_account_class_hash,
             contract_account_class_hash,
+            contract_reader,
         );
 
         Self { starknet_provider, network, kakarot_contract }
     }
 
     /// Returns the result of executing a call on a ethereum address for a given calldata and block
     /// without creating a transaction.
-    pub async fn call(
-        &self,
-        origin: Address,
-        to: Address,
-        calldata: Bytes,
-        block_id: BlockId,
-    ) -> Result<Bytes, EthApiError> {
+    pub async fn call(&self, request: CallRequest, block_id: BlockId) -> Result<Bytes, EthApiError> {
         let starknet_block_id: StarknetBlockId = EthBlockId::new(block_id).try_into()?;
 
+        // unwrap option or return jsonrpc error
+        let to = request
+            .to
+            .ok_or_else(|| EthApiError::MissingParameterError("Missing `to` field in CallRequest".to_string()))?;
         let to: Felt252Wrapper = to.into();
         let to = to.into();
 
+        // Here we check if CallRequest.origin is None, if so, we insert origin = address(0)
+        let origin = request.from.unwrap_or(Address::zero());
         let origin: FieldElement = Felt252Wrapper::from(origin).into();
 
-        let calldata = calldata.to_vec().into_iter().map(FieldElement::from).collect();
-
-        let result = self.kakarot_contract.eth_call(&origin, &to, calldata, &starknet_block_id).await?;
+        let calldata = request
+            .input
+            .data
+            .ok_or_else(|| EthApiError::MissingParameterError("Missing `data` field in CallRequest".to_string()))?;
+        let calldata: Vec<_> = calldata.to_vec().into_iter().map(FieldElement::from).collect();
+
+        let gas_limit: Felt252Wrapper = request.gas.unwrap_or_default().try_into()?;
+        let gas_limit: FieldElement = gas_limit.into();
+
+        let gas_price: Felt252Wrapper = request.gas_price.unwrap_or_default().try_into()?;
+        let gas_price: FieldElement = gas_price.into();
+
+        let value: Felt252Wrapper = request.value.unwrap_or_default().try_into()?;
+        let value: FieldElement = value.into();
+
+        let (_, return_data, success) = self
+            .kakarot_contract
+            .reader
+            .eth_call(&origin, &to, &gas_limit, &gas_price, &value, &calldata.len().into(), &CairoArrayLegacy(calldata))
+            .block_id(starknet_block_id)
+            .call()
+            .await?;
+
+        if success == FieldElement::ZERO {
+            let revert_reason =
+                return_data.0.into_iter().filter_map(|x: FieldElement| u8::try_from(x).ok()).collect::<Vec<_>>();
+            let revert_reason = String::from_utf8(revert_reason).unwrap_or_else(|_| "Unknown".into());
+            return Err(EthApiError::Other(anyhow!("Revert reason: {}", revert_reason)));
```

Comment:
Maybe here we should not return an Error, semantically, maybe semantically it isn't an error

Reply:
I agree, would be inclined to return the revert_reason converted to bytes?

Reply:
I checked `reth`, they return an error if the execution failed, we could add `ExecutionError` to the `EthApiError`
https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/api/call.rs#L69C12-L69C12